File: backend/src/schemas/finance_agent.py
from __future__ import annotations
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel


class FinanceAgentDoctorEarning(BaseModel):
    doctor_name: str
    total_earned: float
    invoices_paid: int


class FinanceAgentAgentCost(BaseModel):
    agent_slug: str
    sessions: int
    estimated_cost: float


class FinanceAgentRecentInvoice(BaseModel):
    patient_name: str
    status: str
    amount: float


class FinanceAgentReport(BaseModel):
    total_revenue: float
    total_expenses: float
    net: float
    outstanding: float
    invoice_count: int
    expense_count: int
    month_label: str = ""
    per_doctor: list[FinanceAgentDoctorEarning] = []
    per_agent_cost: list[FinanceAgentAgentCost] = []
    recent_invoices: list[FinanceAgentRecentInvoice] = []


class AskFinanceAgentRequest(BaseModel):
    question: str
    session_id: UUID | None = None


class AskFinanceAgentResponse(BaseModel):
    session_id: UUID
    answer: str


class FinanceAgentMessage(BaseModel):
    role: Literal["staff", "agent"]
    content: str
    created_at: datetime


class FinanceAgentSessionSummary(BaseModel):
    id: UUID
    title: str
    updated_at: datetime


class FinanceAgentSessionDetail(BaseModel):
    id: UUID
    title: str
    messages: list[FinanceAgentMessage]

File: frontend/src/api/financeAgent.ts
export interface FinanceAgentDoctorEarning {
  doctor_name: string;
  total_earned: number;
  invoices_paid: number;
}

export interface FinanceAgentAgentCost {
  agent_slug: string;
  sessions: number;
  estimated_cost: number;
}

export interface FinanceAgentRecentInvoice {
  patient_name: string;
  status: string;
  amount: number;
}

export interface FinanceAgentReport {
  total_revenue: number;
  total_expenses: number;
  net: number;
  outstanding: number;
  invoice_count: number;
  expense_count: number;
  month_label: string;
  per_doctor: FinanceAgentDoctorEarning[];
  per_agent_cost: FinanceAgentAgentCost[];
  recent_invoices: FinanceAgentRecentInvoice[];
}

export interface AskFinanceAgentResponse {
  session_id: string;
  answer: string;
}

export interface FinanceAgentMessage {
  role: "staff" | "agent";
  content: string;
  created_at: string;
}

export interface FinanceAgentSessionSummary {
  id: string;
  title: string;
  updated_at: string;
}

export interface FinanceAgentSessionDetail {
  id: string;
  title: string;
  messages: FinanceAgentMessage[];
}

type AuthedFetch = <T>(path: string, init?: RequestInit) => Promise<T>;

// Matches backend/src/router/finance_agent/finance_agent_router.py.
export function getFinanceAgentReport(authedFetch: AuthedFetch) {
  return authedFetch<FinanceAgentReport>("/api/v1/finance-agent/report");
}

export function askFinanceAgent(authedFetch: AuthedFetch, question: string, sessionId?: string | null) {
  return authedFetch<AskFinanceAgentResponse>("/api/v1/finance-agent/ask", {
    method: "POST",
    body: JSON.stringify({ question, session_id: sessionId ?? null })
  });
}

export function listFinanceAgentSessions(authedFetch: AuthedFetch) {
  return authedFetch<FinanceAgentSessionSummary[]>("/api/v1/finance-agent/sessions");
}

export function getFinanceAgentSession(authedFetch: AuthedFetch, sessionId: string) {
  return authedFetch<FinanceAgentSessionDetail>(`/api/v1/finance-agent/sessions/${sessionId}`);
}